[이다빈] Mission 5,6: 회원가입 비밀번호 암호화 적용, 로그인 JWT 적용 (2차수정)

thisisempty opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!


**vscode 세팅 때문인지 한줄로 변경이 안됩니다 나머지는 모두 수정했습니다**

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 회원가입 비밀번호 암호화 적용, jwt 적용
- coding convention 관련 수정
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 회원가입 시 패스워드를 암호화 해서 데이터베이스에 저장할 수 있음
- 이메일과 패스워드 사용해서 로그인 할 수 있음
- 사용자의 엑세스 토큰으로 권한을 확인할 수 있음
- 성공시 200에러 반환
- 실패시 해당 에러 반환

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

thisisempty committed:
[ADD]
westagram/settings.py : add 'users' to 'INSTALLED_APPS'
westagram/urls.py : add 'users' APP path

[CREATE]
users/views.py : create new view 'SignInView' and 'SignUpView'
users/urls.py : create new path 'signin' and 'signup'

thisisempty committed:
[MODIFY]
users/models.py : Change the phone column from an IntegerField to aCharfield.

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+from django.views import View
+from users.models import User
+
 # Create your views here.
+
```
주석과 빈 줄은 지워주세요~
```suggestion

```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            email = data['email']
+            password = data['password']
+
+            # Return 400 error when using an existing email.
```
주석 없이도 충분히 가독성 좋은 코드를 작성해주셔서, 주석은 필요 없을 것 같아요~
```suggestion
            # Return 400 error when using an existing email.
```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            password = data['password']
+
+            # Return 400 error when using an existing email.
+            if User.objects.filter(email=email).exists():
+                return JsonResponse({'message': 'EMAIL_ALREADY_EXISTS'}, status=400)
```
자원이 존재하지 않는 경우 status code로 404를 반환해주세요~
```suggestion
            if User.objects.filter(email=email).exists():
                return JsonResponse({'message': 'EMAIL_ALREADY_EXISTS'}, status=400)
```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            regex_email = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
+            regex_password = '^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9]).{8,}'
+
+            # Return message if not satisfied with the validation rules.
```
마찬가지 입니다~
```suggestion
```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            if User.objects.filter(email=email).exists():
+                return JsonResponse({'message': 'EMAIL_ALREADY_EXISTS'}, status=400)
+
+            regex_email = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
+            regex_password = '^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9]).{8,}'
```
정규식을 complie하지 않고 상수로 선언해서 사용하셨네요!  👍 👍 
추가로 상수는 대문자로 작성해주세요.

```suggestion
            REGEX_EMAIL = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
            REGEX_PASSWORD = '^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9]).{8,}'
```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
 from django.shortcuts import render
```
사용하지 않는 모듈은 불필요하게 import 하지 않도록 지워주세요~

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
 from django.shortcuts import render
 
+import json
+import bcrypt
+import re
+import jwt
+from django.http import JsonResponse
+from django.views import View
+from users.models import User
```
import convention에 맞춰주세요.

```suggestion
# bulit-in module
import json
import bcrypt
import re

# 외부 모듈
import jwt
from django.http import JsonResponse
from django.views import View

# 커스텀 모듈
from users.models import User
```

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            return JsonResponse({'message': 'SUCCESS'}, status=201)
+
+        except KeyError:
+            # Return 400 error if email or password is not delivered.
```
지워주세요~

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+            if not re.match(regex_password, password):
+                return JsonResponse({'message': 'INVAILD_PASSWORD'})
+
+            # Store user's password by hash-encrypted.
```
마찬가지로 지워주세요~

lordmyshepherd commented on students/dabeen/westagram/users/views.py:
```diff
@@ -1,3 +1,77 @@
+                user = User.objects.get(email=data['email'])
+
+                if bcrypt.checkpw(data['password'].encode('utf-8'), user.password.encode('utf-8')) == True:
+                    access_token = jwt.encode(
+                        {'id': user.id}, 'secret', algorithm='HS256')
```
```suggestion
                    access_token = jwt.encode({'id': user.id}, 'secret', algorithm='HS256')
```

lordmyshepherd replied:
한 줄로 바꿔주시고 algorithm은 my_settings.py로 옮겨 주시고 `from .settings import ALGORITHM`을 통해서 import해서 가져올 수 있도록 수정해주세요~

thisisempty replied:
다른것 다 수정해서 커밋했는데 충돌이 해결이 안되네요 ㅜㅜ

thisisempty committed:
[modify]
westagram/urls.py : modify conflict-related
wetagram/settings.py : add ALGORITHM
users/models.py : modify confilct-related
users/views.py : modify according to coding convention

thisisempty committed:
migration modified

thisisempty committed:
conflict 관련 커밋

thisisempty committed:
conflict 관련 커밋

kyeongraekim commented on students/dabeen/westagram/users/views.py:
```diff
@@ -43,16 +54,20 @@ class SignInView(View):
+                    {'id': user.id}, SECRET_KEY, algorithm=ALGORITHM)
 
-            return JsonResponse({"message": "SUCCESS"}, status=200)
+                return JsonResponse(jwt.decode(access_token, SECRET_KEY, algorithm=ALGORITHM), status=200)
```
encoding 한 토큰을 다시 decoding 해서 보내면 보안성에 문제가 생길 것 같아 보여요. 로그인시 성공 메세지, 알맞은 상태코드와 함께 인코딩된 access_token 그대로를 리턴할 수 있게 수정해주세요.

thisisempty replied:
수정해서 커밋 했습니다.

kyeongraekim reviewed (changes requested):
컨플릭트 해결완료 확인했습니다. 마지막 한 가지 수정 부분 더 확인해주세요!~

thisisempty committed:
modify access_token return part

thisisempty committed:
modify access_token return part

lordmyshepherd reviewed (approved):
LGTM 👍
